### PR TITLE
Fix 3D Perspective Mode

### PR DESCRIPTION
**Pull request by RobertBColton**
This one is going to be a little bit complicated to explain, but from my testing, `d3d_set_perspective` in GM8.1 and GMSv1.4 should **never** take effect until the next frame/`screen_redraw`. It should also only take effect if we are actually in d3d mode meaning that `d3d_start()` has been called.

The newer GMS manual describes this function as if it affects the "default" room projection because it mentions the view:
https://docs.yoyogames.com/source/dadiospice/002_reference/drawing/drawing%203d/3d%20setup/d3d_set_perspective.html

>This turns on (or off) perspective projection. With a perspective projection instances that have a greater depth will appear smaller, so when the depth is 0 it is equal to the original size and the viewpoint for the camera is placed at a distance above the room. This default distance is equal to the width of the room as that gives a reasonable starting projection, and only instances in front of the camera are drawn. Don't use instances with a depth smaller than 0 (or at least not smaller than -w where w is the width of the room or the view).

I've tested this on several games which make use of the `d3d_set_perspective` function including the High Polygonal 3D Cubes Demo, House Effects Demo, Animation Platform Example (DarkAceZ game). I did not discover any regressions as a result of these changes here. However, I should also mention that I notice no improvements in any game that makes use of this function either (probably due to the generally poor understanding of how it was supposed to work even in GM8.1).

To clarify: `d3d_set_perspective` only affects the **_starting_** projection at the beginning of every frame when in 3D mode

As usual, I have created a test GMK: [perspective-projection-test.zip](https://github.com/enigma-dev/enigma-dev/files/2386685/perspective-projection-test.zip)

In GM8.1, GMSv1.4, and ENIGMA (with this pr applied) the game draws the text inverted. This means that the orthographic projection set immediately after `d3d_start` does not take effect because it is overridden by the perspective projection that GM sets before each draw event. If we move the setting of the orthographic projection into the draw event of each, then the text is drawn normally and not inverted. If we run this test project on ENIGMA master, the text is just always drawn normal without being inverted (incorrect because the perspective projection is never applied).

Another example that proves `d3d_set_perspective` does not have any effect until the **_next_** frame is the following:
```gml
d3d_set_projection_ortho(0, 0, room_width, room_height, 0);
d3d_set_perspective(true); // no effect in 8.1 or GMS
draw_text(0, 0, "ABCDEFGHIJKLMNOP!");
```

### CODE REVIEW

**Comment by RobertBColton** [comment]
@JoshDreamland ok this one is basically good to go. I've gone to great depths to make sure this is correct. I don't think we should worry about this too much considering our existing implementations of the functions were totally broken anyway, they would only ever set a perspective one but never unset it. This is definitely a more correct change that behaves more like GM and doesn't appear to break anything. Please take a look at it, thanks!

File: ENIGMAsystem/SHELL/Graphics_Systems/General/GSscreen.cpp [hunk 1]
```diff
@@ -153,7 +153,10 @@ static inline int draw_tiles()
 
 void clear_view(float x, float y, float w, float h, float angle, bool showcolor)
 {
-  d3d_set_projection_ortho(x, y, w, h, angle);
+  if (enigma::d3dMode && enigma::d3dPerspective)
+    d3d_set_projection_perspective(x, y, w, h, angle);
```
Comment (RobertBColton) [inline ENIGMAsystem/SHELL/Graphics_Systems/General/GSscreen.cpp]: @JoshDreamland this is where I (and GM) makes `d3d_set_perspective` actually take effect.

**Comment by RobertBColton** [comment]
I did this test too which proves that `d3d_start` and `d3d_end` do change the same variable as `d3d_set_perspective`:
```gml
d3d_set_perspective(true);
d3d_start();
d3d_set_perspective(true);
d3d_end();
d3d_set_perspective(true);
```

File: ENIGMAsystem/SHELL/Graphics_Systems/General/GSscreen.cpp (hunk 1 above)
Comment (JoshDreamland) [reply ENIGMAsystem/SHELL/Graphics_Systems/General/GSscreen.cpp]: I might instead suggest storing the default protection in a matrix, although I suppose the parameters might not multiply in cleanly. This approach is therefore fine, but please consider a `draw_clear_projection` method, or maybe `d3d_reset_projection`, or your choice of permutation.

**Comment by RobertBColton** [comment]
Alright, yeah I finally figured out how to get [apitrace](https://github.com/apitrace/apitrace) working with regular old GM. You have to use the x86 build of apitrace on GM's built executables. Doing this I was able to prove definitively that GM resets the viewport and projection matrix every frame

```cpp
// Create Event
d3d_start();
frames = 0;
// Draw Event
draw_line(0, 0, 10, 10);
frames += 1;
if (frames > 5) {
    d3d_set_perspective(false);
}
if (frames > 10) {
    game_end();
}
```

GM8 API Trace

File: ENIGMAsystem/SHELL/Graphics_Systems/General/GSscreen.cpp (hunk 1 above)
Comment (RobertBColton) [reply ENIGMAsystem/SHELL/Graphics_Systems/General/GSscreen.cpp]: That would be a good idea, but we're not exactly set up to do that cleanly yet until I generalize the transforms. I am not entirely sure I understand the suggested functions though.

**Comment by RobertBColton** [comment]
In regards to making clear_view take an index instead of all the parameters, we can't do that because we use this helper function even when views are disabled. When that is the case we use the window region functions to determine the scaled region to clear as the viewport.

**Pull request merged**